File: src/test/java/no/nav/samordning/hendelser/feed/FeedControllerAuthTest.java
package no.nav.samordning.hendelser.feed;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;

import static no.nav.samordning.hendelser.TestTokenHelper.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * Tests authentication/authorization (token validation) of the feed controller API.
 */
@SpringBootTest
@AutoConfigureMockMvc
class FeedControllerAuthTest {

    private static final String ENDPOINT = "/hendelser";
    private static final String AUTH_HEADER_NAME = "Authorization";
    private static final String TPNR_PARAM_NAME = "tpnr";
    private static final String ORG_NUMBER_1 = "0000000000";
    private static final String ORG_NUMBER_2 = "4444444444";
    private static final String TPNR_1 = "1000";
    private static final String TPNR_2 = "4000";

    @Autowired
    private MockMvc mockMvc;

    @Test
    void valid_token_issuer_is_authorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, token(ORG_NUMBER_1, true))
                .param(TPNR_PARAM_NAME, TPNR_1))
                .andExpect(status().isOk());
    }

    @Test
    void valid_orgno_and_tpnr_is_authorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, token(ORG_NUMBER_2, true))
                .param(TPNR_PARAM_NAME, TPNR_2))
                .andExpect(status().isOk());
    }

    @Test
    void expired_token_is_unauthorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, expiredToken(ORG_NUMBER_2))
                .param(TPNR_PARAM_NAME, TPNR_2))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void future_token_is_unauthorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, futureToken(ORG_NUMBER_2))
                .param(TPNR_PARAM_NAME, TPNR_2))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void invalid_token_issuer_is_unauthorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, token(ORG_NUMBER_1, false))
                .param(TPNR_PARAM_NAME, TPNR_1))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void invalid_orgno_is_unauthorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, token("1111111111", false))
                .param(TPNR_PARAM_NAME, TPNR_1))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void invalid_tpnr_is_unauthorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, token(ORG_NUMBER_1, false))
                .param(TPNR_PARAM_NAME, "1235"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void invalid_orgno_and_tpnr_is_unauthorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, token("9999999999", true))
                .param(TPNR_PARAM_NAME, "2000"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void no_token_is_unauthorized() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .param(TPNR_PARAM_NAME, TPNR_2))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void missing_required_parameter_returns_bad_request() throws Exception {
        mockMvc.perform(get(ENDPOINT)
                .header(AUTH_HEADER_NAME, emptyToken())
                .param(TPNR_PARAM_NAME, TPNR_2))
                .andExpect(status().isBadRequest());
    }
}

File: src/test/kotlin/no/nav/samordning/hendelser/feed/FeedControllerAuthTest.kt
package no.nav.samordning.hendelser.feed

import io.zonky.test.db.AutoConfigureEmbeddedDatabase
import no.nav.pensjonsamhandling.maskinporten.validation.test.AutoConfigureMaskinportenValidator
import no.nav.pensjonsamhandling.maskinporten.validation.test.MaskinportenValidatorTokenGenerator
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get

/**
 * Tests authentication/authorization (maskinportenValidatorTokenGenerator) of the feed controller API.
 */
@SpringBootTest
@AutoConfigureMaskinportenValidator
@AutoConfigureMockMvc
@AutoConfigureEmbeddedDatabase(provider = AutoConfigureEmbeddedDatabase.DatabaseProvider.ZONKY)
internal class FeedControllerAuthTest {

    @Autowired
    private lateinit var maskinportenValidatorTokenGenerator: MaskinportenValidatorTokenGenerator

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Test
    fun `Valid request is ok`() {
        mockMvc.get(ENDPOINT) {
            headers {
                setBearerAuth(maskinportenValidatorTokenGenerator.generateToken(SCOPE_SAMORDNING, "889640782").serialize())
            }
            param(TPNR_PARAM_NAME, TPNR_1)
        }.andExpect {
            status { isOk() }
        }
    }

    @Test
    fun `Request failing validation is forbidden`() {
        mockMvc.get(ENDPOINT) {
            headers {
                setBearerAuth(maskinportenValidatorTokenGenerator.generateToken(SCOPE_SAMORDNING, "889640777").serialize())
            }
            param(TPNR_PARAM_NAME, TPNR_1)
        }.andExpect {
            status { isForbidden() }
        }
    }

    @Test
    fun `Request illegal scope is forbidden`() {
        mockMvc.get(ENDPOINT) {
            headers {
                setBearerAuth(maskinportenValidatorTokenGenerator.generateToken("ILLEGAL_SCOPE", "889640782").serialize())
            }
            param(TPNR_PARAM_NAME, TPNR_1)
        }.andExpect {
            status { isForbidden() }
        }
    }

    @Test
    fun `Missing token is unauthorized`() {
        mockMvc.get(ENDPOINT) {
            param(TPNR_PARAM_NAME, TPNR_2)
        }.andExpect {
            status { isUnauthorized() }
        }

    }

    @Test
    fun `Missing parameter is bad request`() {
        mockMvc.get(ENDPOINT) {
            headers {
                setBearerAuth(maskinportenValidatorTokenGenerator.generateToken(SCOPE_SAMORDNING, "889640782").serialize())
            }
        }.andExpect {
            status { isBadRequest() }
        }
    }

    companion object {
        private const val URL_VEDTAK = "/hendelser?tpnr=1000"
        private const val SCOPE_SAMORDNING = "nav:pensjon/v1/samordning"
        private const val ENDPOINT = "/hendelser"
        private const val TPNR_PARAM_NAME = "tpnr"
        private const val TPNR_1 = "1000"
        private const val TPNR_2 = "4000"
    }
}
